src/esp32/single_encoder.rs

```diff
@@ -0,0 +1,341 @@
+use embedded_hal::digital::v2::InputPin;
+
+use super::encoder::PulseStorage;
+use super::pin::PinExt;
+use super::pulse_counter::{get_unit, isr_install, isr_installed, isr_uninstall};
+
+use crate::common::encoder::{
+    Direction, Encoder, EncoderPosition, EncoderPositionType, EncoderSupportedRepresentations,
+    SingleEncoder,
+};
+
+use core::ffi::{c_short, c_ulong};
+use esp_idf_sys as espsys;
+use espsys::pcnt_channel_edge_action_t_PCNT_CHANNEL_EDGE_ACTION_DECREASE as pcnt_count_dec;
+use espsys::pcnt_channel_edge_action_t_PCNT_CHANNEL_EDGE_ACTION_INCREASE as pcnt_count_inc;
+use espsys::pcnt_channel_level_action_t_PCNT_CHANNEL_LEVEL_ACTION_KEEP as pcnt_mode_keep;
+use espsys::pcnt_channel_t_PCNT_CHANNEL_0 as pcnt_channel_0;
+use espsys::pcnt_config_t;
+use espsys::pcnt_evt_type_t_PCNT_EVT_H_LIM as pcnt_evt_h_lim;
+use espsys::pcnt_evt_type_t_PCNT_EVT_L_LIM as pcnt_evt_l_lim;
+use espsys::{esp, EspError, ESP_OK};
+
+use std::collections::BTreeMap;
+use std::sync::atomic::{AtomicI32, Ordering};
+use std::sync::{Arc, Mutex};
+
+use crate::common::status::Status;
+
+// TODO: Make configurable?
+const MAX_GLITCH_MICROSEC: u16 = 1;
+
+// TODO: Move this type to common once we have a single encoder
+// implementation for another board
+pub(crate) type SingleEncoderType = Arc<Mutex<dyn SingleEncoder>>;
+
+pub struct Esp32SingleEncoder {
+    pulse_counter: Box<PulseStorage>,
+    config: pcnt_config_t,
+    dir: Direction,
+}
+
+impl Esp32SingleEncoder {
+    pub fn new(encoder_pin: impl InputPin + PinExt, dir_flip: bool) -> anyhow::Result<Self> {
+        let unit = get_unit()?;
+        println!("unit received: {:?}", unit);
+        let pcnt = Box::new(PulseStorage {
+            acc: Arc::new(AtomicI32::new(0)),
+            unit,
+        });
+        let mut enc = Esp32SingleEncoder {
+            pulse_counter: pcnt,
+            config: pcnt_config_t {
+                pulse_gpio_num: encoder_pin.pin(),
+                ctrl_gpio_num: -1,
```

Comment:
did you try to set it to the pin dir or one of the A/B pin?

Reply:
The single encoder implementation in RDK simply counts events received by interrupt, so I thought it was safer to emulate that behavior as closely as possible.

Reply:
I agree but if we can leverage the pcnt as much as possible and avoid code it would be best. 

With the control pin you can set the counter to increase or decrease based on the pin level so if it's attached to the dir pin it can make the encoder work like the Quadrature encoder. It should also work with A/B. 

Otherwise you can configure the PCNT to count backward or forward programmatically.

Reply:
Wait, so why do we have a single encoder implementation even in RDK? Couldn't we have just used the incremental encoder implementation and supply it with the dir or A/B pin?